### ISSUE TITLE
Assert that some tests must not be skipped under certain CI configurations

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#4440]
It seems pretty easy to accidentally misconfigure one of our CI environments, and consequently fail to run a test. It would be great if we could somehow either:

1. Assert that all tests eventually get run under some configuration (not sure how to do this)
2. Assert that in a given CI environment, some test must not be skipped

That way, an env misconfiguration turns into an error, rather than just some more tests start getting skipped.

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra @mruberry @VitalyFedyunin

**Comment by ezyang** [comment]
Bumping this issue. This still applies, and would have caught some recent problems.

**Comment by ezyang** [comment]
This still applies.

**Comment by mattip** [comment]
I can imagine two strategies:
- try to proactively parse the test runners in the `.jenkins/pytorch/*test.sh` and `.jenkins/pytorch/win-test-helpers/test_python_nn.bat`
- try to analyze the uploaded test results (circleci uploads to `/home/circleci/project/test-reports/` - is there a way to access that?)

In any case, is it safe to say the set of tests to be run is the one in `run_tests.TEST`?

**Comment by mattip** [comment]
There are also two jobs on Jenkins, a build and a test, the test one reports
```
[GoogleTest-1.6] - 151 test report file(s) were found with the pattern \
  'caffe2_tests/cpp/*.xml' relative to \
  '/var/lib/jenkins/workspace/caffe2-builds/py3.6-devtoolset7-rocmrpm-centos7-test' \
  for the testing framework 'GoogleTest-1.6'.
[xUnit] [INFO] - Processing JUnit
[xUnit] [INFO] - [JUnit] - 1 test report file(s) were found with the pattern \
  'caffe2_tests/python/*.xml' relative to \
  '/var/lib/jenkins/workspace/caffe2-builds/py3.6-devtoolset7-rocmrpm-centos7-test'\
   for the testing framework 'JUnit'.
```

Are those also relevant for this issue?

**Comment by ezyang** [comment]
> In any case, is it safe to say the set of tests to be run is the one in run_tests.TEST?

I don't think so, actually; I think sometimes people add more test files to `test/` and forget to update `run_tests.py` I think this is a more rare occurrence.

> There are also two jobs on Jenkins, a build and a test, the test one reports

Yeah, the test reports are relevant; they could be used to aggregate information about what tests actually ran versus not.

**Comment by mattip** [comment]
It seems there already is a [Dr CI](https://github.com/kostmo/circleci-failure-tracker) integration that aggregates CI runs. I wonder if we could extend that to check the report logs for all the test CI runs, and compare that to a list of "all possible tests" - which would also need to be generated somehow

**Comment by rgommers** [comment]
@kostmo any opinion on using Dr CI for this?

**Comment by ezyang** [comment]
@kostmo is on paternity leave right now. With the folks we have around right now, it might not be so easy to add this feature to Dr CI.
